File: code/EA.nim
import Individual

import std/random

proc generateRandomOffspring(x: Individual, p: float): Individual =
    #[ generate copy of x flipping each bit independently with probability p ]#
    let n = x.size
    let y = newIndividual(n)
    for idx in countup(0, n-1):
        var xi = get(x, idx)
        let randVar = rand(1.0)
        if (xi == 0 and randVar < p) or (xi == 1 and randVar >= p):
            set(y, idx)
    return y

proc EvolutionaryAlgorithm*(f: proc (x: Individual): int, n: int): Individual =
    #[ (1+1) Evolutionary Algorithm ]#

    # Call randomize() once to initialize the default random number generator.
    # If this is not called, the same results will occur every time these
    # examples are run.
    randomize()

    var t = 0
    var Pt = generateRandomOffspring(newIndividual(n), 0.5) #random initial solution
    
    while t < 100:

        let y = generateRandomOffspring(Pt, 1/n)
        var Qt: array[2, Individual]

        if(f(y) < f(Pt)): #we sort population and offspring with respect to f
            Qt = [y, Pt]
        else:
            Qt = [Pt, y]

        t += 1
        Pt = Qt[1] #we pick solution in population that maximises f
    return Pt

File: code/EA.py
from Individual import Individual
import numpy as np
import random
import heapq

def generateRandomOffspring(x, p):
    """
    Generate a copy of x flipping each bit independently with probability p
    """
    n = x.size
    y = Individual(n, x.t + 1) #IMPORTANT: increment time counter!
    for idx in range(n):
        xi = x.get(idx)
        rand_var = np.random.uniform(0, 1)
        #bit idx in y is 1 if and only if
        mutated_to_one = (rand_var < p and xi == 0) #mutated from 0 to 1 (with probability p)
        stayed_one = (rand_var >= p and xi == 1) #did not mutate, was already 1 (probability 1-p)
        if (mutated_to_one or stayed_one): 
            y.set(idx)
    return y

def EvolutionaryAlgorithm(f, n):
    """
    (1+1) Evolutionary Algorithm
    """
    t = 0
    Pt = generateRandomOffspring(Individual(n, 0), 0.5)  # random initial solution (t=0)

    while Pt.count() < n:
        y = generateRandomOffspring(Pt, 1 / n)
        if f(y) >= f(Pt):  # IMPORTANT: Pick solution that maximises f, in case of tie CHOOSE OFFSPRING!!!
            Pt = y
        t += 1

    return Pt

def EvolutionaryAlgorithm2(f, n, mu):
    """
    (mu + 1) Evolutionary Algorithm
    """
    t = 0
    Pt = []
    most_fit_individual = (0, Individual(n, -1)) #used to store solution
    
    for _ in range(mu):# we must create a random initial population of size mu
        individual = generateRandomOffspring(Individual(n, 0), 0.5)
        fitness = f(individual)
        pair_fitness_individual = (fitness, individual) #we create pair (fitness, individual)

        if(fitness >= most_fit_individual[0]):
            most_fit_individual = pair_fitness_individual

        Pt.append(pair_fitness_individual)

    heapq.heapify(Pt) # we transform population into a priority queue

    while True:

        if(most_fit_individual[1].count() == n):
            return most_fit_individual[1]
            
        offspring = generateRandomOffspring(random.choice(Pt)[1], 1 / n)
        fitness = f(offspring)
        pair_fitness_offspring = (fitness, offspring)

        if(fitness >= most_fit_individual[0]):
            most_fit_individual = pair_fitness_offspring

        heapq.heappushpop(Pt, pair_fitness_offspring) #we add offspring to the priority queue and pop least fit individual
        t += 1